### ISSUE TITLE
[Inference] Speculative decoding for 32B

### ISSUE THREAD

**Issue by wallscaler** [wallscaler/baobab-tts#9]
# Issue #9: [Inference] Speculative Decoding for 32B

## Labels
`inference`, `optimization`, `priority-high`, `phase-4`

## Objective

Implement VADUSA-style speculative decoding to achieve 5x faster inference for the 32B model.

## Background

From [VADUSA](https://arxiv.org/abs/2410.21951):

> "Up to 5× faster inference than state-of-the-art autoregressive TTS models without sacrificing quality"

Speculative decoding uses a lightweight draft model to predict future tokens, then verifies them in parallel with the main model.

### Architecture

```
Standard Autoregressive (Slow):
┌─────────────────────────────────────────────┐
│ Step 1: [Token 1] ──────────────────────────│
│ Step 2: [Token 1, Token 2] ─────────────────│
│ Step 3: [Token 1, Token 2, Token 3] ────────│
│ Step 4: [Token 1, Token 2, Token 3, 4] ─────│
│ ...                                         │
└─────────────────────────────────────────────┘
Time: O(n) forward passes

Speculative Decoding (Fast):
┌─────────────────────────────────────────────┐
│ Draft:  [1, 2?, 3?, 4?, 5?] (in parallel)   │
│ Verify: [1✓, 2✓, 3✓, 4✗, 5?] (in parallel)  │
│ Accept: [1, 2, 3]                           │
│ Reject: Regenerate from 4                   │
└─────────────────────────────────────────────┘
Time: O(n/k) forward passes, where k is acceptance rate
```

### Draft Heads

Add lightweight prediction heads to the main model:

```python
class DraftHeads(nn.Module):
    """Lightweight heads for speculative token prediction."""

    def __init__(
        self,
        hidden_size: int,
        vocab_size: int,
        num_heads: int = 4,  # Predict 4 tokens ahead
    ):
        super().__init__()
        self.heads = nn.ModuleList([
            nn.Sequential(
                nn.Linear(hidden_size, hidden_size // 4),
                nn.ReLU(),
                nn.Linear(hidden_size // 4, vocab_size),
            )
            for _ in range(num_heads)
        ])

    def forward(self, hidden_states: Tensor) -> List[Tensor]:
        """Predict next K tokens from current hidden state."""
        return [head(hidden_states) for head in self.heads]
```

### Tolerance Mechanism

VADUSA introduces tolerance for near-matches:

```python
class SpeculativeDecoder:
    """Speculative decoding with tolerance."""

    def __init__(
        self,
        model: BaobabModel,
        draft_heads: DraftHeads,
        tolerance: float = 0.1,  # Accept if within top 10%
    ):
        self.model = model
        self.draft = draft_heads
        self.tolerance = tolerance

    def generate(
        self,
        text: str,
        max_length: int = 4096,
    ) -> Tensor:
        """Generate with speculative decoding."""
        tokens = self.model.encode_text(text)

        while len(tokens) < max_length:
            # 1. Get hidden state from main model
            hidden = self.model.forward_hidden(tokens)

            # 2. Draft K future tokens
            draft_logits = self.draft(hidden[-1])
            draft_tokens = [logits.argmax(-1) for logits in draft_logits]

            # 3. Verify all at once with main model
            candidate = tokens + draft_tokens
            verify_logits = self.model.forward(candidate)

            # 4. Accept/reject with tolerance
            accepted = []
            for i, (draft_tok, verify_logit) in enumerate(
                zip(draft_tokens, verify_logits[-len(draft_tokens):])
            ):
                # Check if draft token is in top-p
                probs = F.softmax(verify_logit, dim=-1)
                top_p_tokens = get_top_p_tokens(probs, p=1-self.tolerance)

                if draft_tok in top_p_tokens:
                    accepted.append(draft_tok)
                else:
                    # Reject and sample from verify distribution
                    accepted.append(sample(probs))
                    break  # Stop accepting after first rejection

            tokens = tokens + accepted

            # Check for EOS
            if tokens[-1] == self.model.eos_token_id:
                break

        return tokens
```

### Training Draft Heads

```python
def train_draft_heads(
    model: BaobabModel,
    draft_heads: DraftHeads,
    train_data: DataLoader,
    num_epochs: int = 1,
):
    """Train draft heads on model's own generations."""
    optimizer = AdamW(draft_heads.parameters(), lr=1e-4)

    for epoch in range(num_epochs):
        for batch in train_data:
            # Get hidden states and target tokens
            with torch.no_grad():
                hidden = model.forward_hidden(batch.tokens)
                targets = batch.tokens[:, 1:]  # Shifted

            # Predict from each position
            losses = []
            for i, head in enumerate(draft_heads.heads):
                logits = head(hidden[:, :-i-1])  # Predict i+1 ahead
                target = targets[:, i:]
                loss = F.cross_entropy(logits, target)
                losses.append(loss)

            total_loss = sum(losses) / len(losses)
            total_loss.backward()
            optimizer.step()
            optimizer.zero_grad()
```

## Configuration

```yaml
inference:
  speculative:
    enabled: true
    num_draft_heads: 4
    tolerance: 0.1
    max_draft_length: 8  # Maximum tokens to draft at once

  draft_head:
    hidden_size: 1024  # 1/4 of main model
    num_layers: 2
```

## Expected Results

| Model | Baseline Speed | With Speculative | Speedup |
|-------|----------------|------------------|---------|
| TTS-1 | 1× | 5× | 5× |
| Spark-TTS | 1× | 4.2× | 4.2× |
| **Baobab 32B** | 1× | **~4×** | **4×** |

## Files to Create

- [ ] `src/inference/__init__.py`
- [ ] `src/inference/draft_heads.py`
- [ ] `src/inference/speculative.py`
- [ ] `tests/test_speculative.py`
- [ ] `configs/inference/speculative.yaml`
- [ ] `scripts/benchmark_inference.py`

## Acceptance Criteria

- [ ] Draft heads train successfully
- [ ] Speculative decoding works
- [ ] Output quality matches standard decoding
- [ ] 4x+ speedup achieved
- [ ] Works with streaming
- [ ] Memory overhead acceptable

## Testing Plan

1. **Quality test**: Compare outputs vs standard decoding
2. **Speed test**: Measure wall-clock speedup
3. **Acceptance rate**: Monitor % of drafts accepted
4. **Memory test**: Profile GPU memory with draft heads
5. **Streaming test**: Verify works with chunked output

## References

- [VADUSA](https://arxiv.org/abs/2410.21951) - Speculative decoding for TTS
- [Speculative Decoding](https://arxiv.org/abs/2211.17192) - Original paper
- [Medusa](https://arxiv.org/abs/2401.10774) - Multi-head speculative
